python/perf-kernels/flash-attention.py

```diff
@@ -203,44 +206,39 @@ def _attn_fwd_inner(
         lo, hi = 0, seqlen_k
     # loop over k, v and update accumulator
     for start_n in range(lo, hi, BLOCK_N):
+        '''
         if STAGE == 1 or STAGE == 3:
             start_n = tl.multiple_of(start_n, BLOCK_N)
+        '''
         # For padded blocks, we will overrun the tensor size if
         # we load all BLOCK_N. For others, the blocks are all within range.
-        if PADDED_BLOCK:
-            if PADDED_HEAD:
-                k = tl.load(K_block_ptr, boundary_check=(1,0), padding_option="zero")
-            else:
-                k = tl.load(K_block_ptr, boundary_check=(1,), padding_option="zero")
+        if (PADDED_BLOCK or STAGE == 2) or PADDED_HEAD:
+            k = tl.load(K_block_ptr, boundary_check=(1,0), padding_option="zero")
         else:
-            if PADDED_HEAD:
-                k = tl.load(K_block_ptr, boundary_check=(0,), padding_option="zero")
-            else:
-                k = tl.load(K_block_ptr)
+            k = tl.load(K_block_ptr, boundary_check=(0,), padding_option="zero")
         if PRE_LOAD_V:
-            if PADDED_BLOCK:
-                if PADDED_HEAD:
-                    v = tl.load(V_block_ptr, boundary_check=(0,1), padding_option="zero")
-                else:
-                    v = tl.load(V_block_ptr, boundary_check=(0,), padding_option="zero")
+            if (PADDED_BLOCK or STAGE == 2) or PADDED_HEAD:
+                v = tl.load(V_block_ptr, boundary_check=(0,1), padding_option="zero")
             else:
-                if PADDED_HEAD:
-                    v = tl.load(V_block_ptr, boundary_check=(1,), padding_option="zero")
-                else:
-                    v = tl.load(V_block_ptr)
```

Comment:
I haven't figured out the optimal way to set the `boundary_check`. This is just for safe, but for most cases, esp. smaller seqlen_q/k `boundary_check=(1,)` is required.

Reply:
I provided an example of how to do this already. Please use that.

https://github.com/ROCm/triton/blob/varlen_causal/python/perf-kernels/flash-attention.py#L132

I don't think readability is a good reason to tank performance, which this will do.

Reply:
> I don't think readability is a good reason to tank performance, which this will do.

The real disaster is, well, even this piece of code is not enough to pass all tests.
More boundary checks are needed: https://github.com/ROCm/aotriton/commit/b892693a3217070c387f7e1f4c1adb2d9e507949

Performance is the thing we need to worry about only after getting all tests passed.

Reply:
> If you have tests failing due to bugs, please either fix the bug and send a PR

This change IS the bug fix.

Reply:
No. This is a heavy hammer you're using to ensure there are no out of bounds accesses, ever, irrespective of whether we are aligned to a multiple of BLOCK_N or not. 

There should be no reason to see OOB accesses when V is a multiple of BLOCK_N. If you are seeing OOB accesses in that case, then the bug is elsewhere, not here and THAT is the place to fix it. This is just covering up the real bug, which still exists.

Reply:
> There should be no reason to see OOB accesses when V is a multiple of BLOCK_N

V is not Multiple of BLOCK_N. This IS the parity problem with CUDA that we want to address.

Reply:
If V is not a multiple of BLOCK_N, PADDED_BLOCK is True, and you should use PADDED_BLOCK to mask, which is already handled by the `if` path. You have even done it correctly below on line 274 when `not IS_PRE_LOAD_V`.

Reply:
> you should use PADDED_BLOCK to mask, which is already handled by the `if` path

PADDED_BLOCK is False unconditionally in the `if STAGE & 2` branch.

> You have even done it correctly below on line 274 when `not IS_PRE_LOAD_V`.

That line is exactly WRONG and causes segfaults on some give input sizes.

Reply:
Yes. Because `STAGE & 2` implies we have causal masking and as stated at the top of this kernel in the supported features section:

```
3) Any sequence lengths without padding (currently fwd kernel only, **no causal masking**)
```

This is what my [PR519](https://github.com/ROCm/triton/pull/519) is fixing. So, again, there is no "bug" here. The kernel simply isn't written for `STAGE & 2 and PADDED_BLOCK` at this time, before my PR519 is merged.